pkg/reconciler/apis/apiexportendpointslice/apiexportendpointslice_controller.go

```diff
@@ -0,0 +1,309 @@
+		AddFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSlice(obj)
+		},
+		UpdateFunc: func(_, newObj interface{}) {
+			c.enqueueAPIExportEndpointSlice(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSlice(obj)
+		},
+	})
+
+	globalAPIExportClusterInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSlicesForAPIExport(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSlicesForAPIExport(obj)
+		},
+	})
+
+	globalShardClusterInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.enqueueAllAPIExportEndpointSlices(obj)
+			},
+			UpdateFunc: func(_, newObj interface{}) {
+				c.enqueueAllAPIExportEndpointSlices(newObj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.enqueueAllAPIExportEndpointSlices(obj)
+			},
+		},
+	)
+
+	return c, nil
+}
+
+type APIExportEndpointSlice = apisv1alpha1.APIExportEndpointSlice
+type APIExportEndpointSliceSpec = apisv1alpha1.APIExportEndpointSliceSpec
+type APIExportEndpointSliceStatus = apisv1alpha1.APIExportEndpointSliceStatus
+type Patcher = apisv1alpha1client.APIExportEndpointSliceInterface
+type Resource = committer.Resource[*APIExportEndpointSliceSpec, *APIExportEndpointSliceStatus]
+type CommitFunc = func(context.Context, *Resource, *Resource) error
+
+// controller reconciles APIExportEndpointSlices. It ensures that the shard endpoints are populated
+// in the status of every APIExportEndpointSlices.
+type controller struct {
+	queue workqueue.RateLimitingInterface
+
+	listShards                            func() ([]*corev1alpha1.Shard, error)
+	listAPIExportEndpointSlices           func() ([]*apisv1alpha1.APIExportEndpointSlice, error)
+	getAPIExportEndpointSlice             func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExportEndpointSlice, error)
+	getAPIExport                          func(path logicalcluster.Path, name string) (*apisv1alpha1.APIExport, error)
+	getAPIExportEndpointSlicesByAPIExport func(key string) ([]interface{}, error)
+
+	commit CommitFunc
+}
+
+// enqueueAPIExportEndpointSlice enqueues an APIExportEndpointSlice.
+func (c *controller) enqueueAPIExportEndpointSlice(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	logger := logging.WithQueueKey(logging.WithReconciler(klog.Background(), ControllerName), key)
+	logger.V(4).Info("queueing APIExportEndpointSlice")
+	c.queue.Add(key)
+}
+
+// enqueueAPIExportEndpointSlicesForAPIExport enqueues APIExportEndpointSlices referencing a specific APIExport.
+func (c *controller) enqueueAPIExportEndpointSlicesForAPIExport(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
```

Comment:
change to:
- read the `core.LogicalClusterPathAnnotationKey` annotation for the given APIExport it will give you a path.
- combine the path with the name to get matching keys of `APIExportEndpointSlices` (`path.Join(export.Name)`)
- use the keys to get `APIExportEndpointSlices` objects use `GetByKey` method on the indexer

Reply:
even this does not work. Refefence can be by the canonical path or by logical cluster name. I.e., an export is known under both and `getAPIExportEndpointSlicesByAPIExport` should better take an `*APIExport` as parameter and hide this logic.

Reply:
how can we find a corresponding `APIExportEndpointSlices` for an `APIExport` then?

Reply:
Find those referencing by canonical path (annotation), and then those referencing by cluster name. Return the union.

Reply:
oh, ic, so we miss one more point
... 
- get matching keys of `APIExportEndpointSlices` for `logicalcluster.From(export).Path().Join(export.Name)`

Reply:
two lookup now. done